### PR TITLE
Issue 42 Initial check-in (#1321).

### PR DESCRIPTION
**Pull request by thefazzer**
**Issue 42 Initial check-in (#1321).**

- CT is at 0. 10 headers are presented.
- 10 blocks are downloaded. CT advances to 5.
- Make sure that UnconsumedBlocksDataBytes is equal to
- the sum of serialized sizes of last five blocks.
- CT advances to 10.  
- Make sure UnconsumedBlocksDataBytes is 0.

Preferred: @Aprogiena @noescape00

### CODE REVIEW

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 1]
```diff
@@ -1652,5 +1652,59 @@ public void ChainWith10DataAvailableHeaders_PartialValidationSucceededOnTheFirst
...
+            foreach (ChainedHeader chainedHeader in connectNewHeadersResult.ToHashArray())
+            {
+                chainedHeaderTree.BlockDataDownloaded(chainedHeader, commonChainTip.FindAncestorOrSelf(chainedHeader).Block);
+            }
+
+            // CT advances to 5.
+            chainedHeaderTree.ConsensusTipChanged(chainedHeaderTo.GetAncestor(5));
+
+            int serializedSizeOfBlocks = 0;
+            var lastFive = connectNewHeadersResult.ToHashArray().TakeLast(5);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: unvar

**Comment by Aprogiena** [comment]
Looks good, consider partial validation for each block after download.

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 2]
```diff
@@ -1652,5 +1652,59 @@ public void ChainWith10DataAvailableHeaders_PartialValidationSucceededOnTheFirst
...
+            foreach (ChainedHeader chainedHeader in lastFive)
+            {
+                serializedSizeOfBlocks += chainedHeader.Block.GetSerializedSize();
+            }
+
+            // UnconsumedBlocksDataBytes is equal to the sum of serialized sizes of last five blocks.
+            Assert.Equal(serializedSizeOfBlocks, chainedHeaderTree.UnconsumedBlocksDataBytes);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Maybe also assert the value is not zero, just prevent having some trivial bug which would make the test empty.

**Comment by noescape00** [comment]
Conflicts

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 3]
```diff
@@ -1652,5 +1652,59 @@ public void ChainWith10DataAvailableHeaders_PartialValidationSucceededOnTheFirst
...
+            ConnectNewHeadersResult connectNewHeadersResult = chainedHeaderTree.ConnectNewHeaders(1, listOfChainBlockHeaders);
+            ChainedHeader chainedHeaderTo = connectNewHeadersResult.DownloadTo;
+            chainedHeaderTo.HashBlock.Should().Be(commonChainTip.HashBlock);
+
+            Assert.True(connectNewHeadersResult.HaveBlockDataAvailabilityStateOf(BlockDataAvailabilityState.BlockRequired));
+
+            foreach (ChainedHeader chainedHeader in connectNewHeadersResult.ToHashArray())
+            {
+                chainedHeaderTree.BlockDataDownloaded(chainedHeader, commonChainTip.FindAncestorOrSelf(chainedHeader).Block);
+            }
```
[the remark this answers is not in the document]
Comment (thefazzer) [reply src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Done.

**Comment by thefazzer** [comment]
Now clear of conflicts.

**Pull request merged**